README.md
# get-time-between

> Measure the amount of time during work hours between two dates.

[![NPM version](http://img.shields.io/npm/v/get-time-between.svg?style=flat-square)](https://www.npmjs.com/package/get-time-between)
[![NPM downloads](http://img.shields.io/npm/dm/get-time-between.svg?style=flat-square)](https://www.npmjs.com/package/get-time-between)
[![Build Status](http://img.shields.io/travis/JamieMason/get-time-between/master.svg?style=flat-square)](https://travis-ci.org/JamieMason/get-time-between)
[![Maintainability](https://api.codeclimate.com/v1/badges/80461b911c6c624194a2/maintainability)](https://codeclimate.com/github/JamieMason/get-time-between/maintainability)
[![Follow JamieMason on GitHub](https://img.shields.io/github/followers/JamieMason.svg?style=social&label=Follow)](https://github.com/JamieMason)
[![Follow fold_left on Twitter](https://img.shields.io/twitter/follow/fold_left.svg?style=social&label=Follow)](https://twitter.com/fold_left)

## 🌩 Installation

```
npm install --save get-time-between
```

## 📝 API

Calculate the number of millseconds during working hours between two dates. The
result of this can be formatted as you wish, with libraries such as
[`pretty-ms`](https://github.com/sindresorhus/pretty-ms) for example.

All methods take an optional 3rd argument containing the following options,
shown here with their default values:

```js
import { getTimeBetween } from 'get-time-between';

const fridayFivePm = new Date('May 17 2019 17:00:00');
const mondayElevenAm = new Date('May 20 2017 11:00:00');

const timeBetween = getTimeBetween(fridayFivePm, mondayElevenAm, {
  dailyEnd: [18, 0, 0], // [hour, minute, second] of the end of included days
  dailyStart: [10, 0, 0], // [hour, minute, second] of the start of included days
  excludedDays: [6, 0] // excluded days of the week (0-6 starting Sunday)
});

console.log(timeBetween);
// => 7200000
```

## 👏🏻 Credits

This project is a fork of https://github.com/tal/time-between by
[Tal Atlas](https://github.com/tal).

## 🙋🏾‍♀️ Getting Help

- Get help with issues by creating a
  [Bug Report](https://github.com/JamieMason/get-time-between/issues/new?template=bug_report.md).
- Discuss ideas by opening a
  [Feature Request](https://github.com/JamieMason/get-time-between/issues/new?template=feature_request.md).


CHANGELOG.md
## [2.0.1](https://github.com/JamieMason/get-time-between/compare/2.0.0...2.0.1) (2019-08-04)


### Bug Fixes

* **npm:** update dependencies ([2380f3c](https://github.com/JamieMason/get-time-between/commit/2380f3c))



# [2.0.0](https://github.com/JamieMason/get-time-between/compare/53eb8eb...2.0.0) (2019-06-04)


### Features

* add getTimeBetween method ([53eb8eb](https://github.com/JamieMason/get-time-between/commit/53eb8eb))
* remove getHoursBetween ([cdbe908](https://github.com/JamieMason/get-time-between/commit/cdbe908))
* rename to get-time-between ([596bfad](https://github.com/JamieMason/get-time-between/commit/596bfad))


### BREAKING CHANGES

* Install and import from get-time-between
* Default Export removed
+ Named Export getHoursBetween removed
+ dailyEnd option changed from number to [number, number, number]
+ dailyStart option changed from number to [number, number, number]





time-between.spec.js
var timeBetween = require('./time-between')

describe('timeBetween', () => {
  describe('simple, durring week', () => {
    var start = new Date('March 1 2017 12:00:00')
    var end = new Date('March 2 2017 12:00:00')

    it('should work for single day, overnight', () => {
      var hours = timeBetween(start, end)
      expect(hours).toEqual(8)
    })

    it('should work regardless of order', () => {
      var hours = timeBetween(end, start)
      expect(hours).toEqual(8)
    })

    it('should work for single day', () => {
      var start = new Date('March 2 2017 00:00:00')
      var end = new Date('March 2 2017 23:59:59')

      var hours = timeBetween(start, end)
      expect(hours).toEqual(8)
    })
  })

  describe('over weekend', () => {
    it('should work for single day', () => {
      var start = new Date('March 3 2017 12:00:00')
      var end = new Date('March 6 2017 12:00:00')

      var hours = timeBetween(start, end)
      expect(hours).toEqual(8)
    })

    it('should work for half way into weekend', () => {
      var start = new Date('March 3 2017 12:00:00')
      var end = new Date('March 5 2017 12:00:00')

      var hours = timeBetween(start, end)
      expect(hours).toEqual(6)
    })

    it('should work for a week', () => {
      var start = new Date('March 3 2017 12:00:00')
      var end = new Date('March 10 2017 12:00:00')

      var hours = timeBetween(start, end)
      expect(hours).toEqual(40)
    })

    it('should work for second weekend', () => {
      var start = new Date('March 3 2017 12:00:00')
      var end = new Date('March 11 2017 12:00:00')

      var hours = timeBetween(start, end)
      expect(hours).toEqual(46)
    })

    it('should work for two weekends', () => {
      var start = new Date('March 3 2017 12:00:00')
      var end = new Date('March 13 2017 12:00:00')

      var hours = timeBetween(start, end)
      expect(hours).toEqual(48)
    })
  })

  describe('specific bugs', () => {
    it('should work for this odd error case was getting -4', () => {
      var start = new Date('2017-02-18T03:30:00.000Z')
      var end = new Date('2017-02-18T03:30:00.000Z')

      var hours = timeBetween(start, end)
      expect(hours).toEqual(0)
    })

    it('should work for this odd error case was getting -1', () => {
      var start = new Date('2017-02-13T22:23:19.000Z')
      var end = new Date('2017-02-14T13:51:48.000Z')

      var hours = timeBetween(start, end)
      expect(hours).toEqual(1)
    })

    it('should work for this odd error case was getting -2', () => {
      var start = new Date('2017-02-14T13:51:48.000Z')
      var end = new Date('2017-02-14T13:51:48.000Z')

      var hours = timeBetween(start, end)
      expect(hours).toEqual(0)
    })
  })
})


src/time-between.spec.ts
import { getTimeBetween } from './time-between';

declare global {
  namespace jest {
    interface Matchers<R> {
      toHaveTimeBetween(expectedMsecs: number): R;
    }
  }
}

expect.extend({
  toHaveTimeBetween([start, end], expectedMsecs: number) {
    const result = getTimeBetween(new Date(start), new Date(end));
    const reverseResult = getTimeBetween(new Date(end), new Date(start));
    const pass = result === expectedMsecs && reverseResult === expectedMsecs;
    return {
      message: () => {
        let msg = '';
        msg += pass ? 'did not expect ' : 'expected ';
        msg += `${expectedMsecs} working milliseconds between "${start}" and "${end}" `;
        msg += `but instead received ${result}`;
        return msg;
      },
      pass
    };
  }
});

const hoursToMs = (n: number) => n * 3600000;
const minsToMs = (n: number) => n * 60000;
const secsToMs = (n: number) => n * 1000;

const getInvalidTimeOptions = (dailyStartOrEnd: string) => [
  { [dailyStartOrEnd]: null },
  { [dailyStartOrEnd]: [] },
  { [dailyStartOrEnd]: [11] },
  { [dailyStartOrEnd]: [11, 0] },
  { [dailyStartOrEnd]: ['11', '00', '00'] },
  { [dailyStartOrEnd]: ['11:00:00'] },
  { [dailyStartOrEnd]: [11, 0, 0, 0] },
  { [dailyStartOrEnd]: [24, 0, 0] },
  { [dailyStartOrEnd]: [0, 60, 0] },
  { [dailyStartOrEnd]: [0, 0, 60] },
  { [dailyStartOrEnd]: [-1, 0, 0] },
  { [dailyStartOrEnd]: [0, -1, 0] },
  { [dailyStartOrEnd]: [0, 0, -1] }
];

[
  ...getInvalidTimeOptions('dailyEnd'),
  ...getInvalidTimeOptions('dailyStart'),
  { excludedDays: null },
  { excludedDays: [null] },
  { excludedDays: ['4'] },
  { excludedDays: [-1] },
  { excludedDays: [7] }
].forEach((opts: any) => {
  it(`should throw when invalid options ${JSON.stringify(
    opts
  )} are provided`, () => {
    expect(() => {
      getTimeBetween(new Date(), new Date(), opts);
    }).toThrow();
  });
});

describe('during week', () => {
  it('should work for single day, overnight', () => {
    const start = 'March 1 2017 12:00:00';
    const end = 'March 2 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(8));
  });

  it('should work regardless of order', () => {
    const start = 'March 1 2017 12:00:00';
    const end = 'March 2 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(8));
  });

  it('should work for single day', () => {
    const start = 'March 2 2017 00:00:00';
    const end = 'March 2 2017 23:59:59';
    expect([start, end]).toHaveTimeBetween(hoursToMs(8));
  });
});

describe('over weekend', () => {
  it('should work for single day', () => {
    const start = 'March 3 2017 12:00:00';
    const end = 'March 6 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(8));
  });

  it('should work for half way into weekend', () => {
    const start = 'March 3 2017 12:00:00';
    const end = 'March 5 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(6));
  });

  it('should work for a week', () => {
    const start = 'March 3 2017 12:00:00';
    const end = 'March 10 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(40));
  });

  it('should work for second weekend', () => {
    const start = 'March 3 2017 12:00:00';
    const end = 'March 11 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(46));
  });

  it('should work for two weekends', () => {
    const start = 'March 3 2017 12:00:00';
    const end = 'March 13 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(48));
  });
});

describe('specific bugs', () => {
  it('should work for this odd error case was getting -4', () => {
    const start = '2017-02-18T03:30:00.000Z';
    const end = '2017-02-18T03:30:00.000Z';
    expect([start, end]).toHaveTimeBetween(0);
  });

  it('should work for this odd error case was getting -1', () => {
    const start = '2017-02-13T22:23:19.000Z';
    const end = '2017-02-14T13:51:48.000Z';
    expect([start, end]).toHaveTimeBetween(
      hoursToMs(3) + minsToMs(51) + secsToMs(48)
    );
  });

  it('should work for this odd error case was getting -2', () => {
    const start = '2017-02-14T13:51:48.000Z';
    const end = '2017-02-14T13:51:48.000Z';
    expect([start, end]).toHaveTimeBetween(0);
  });
});


src/get-time-between.spec.ts
import { getTimeBetween } from './get-time-between';

declare global {
  namespace jest {
    interface Matchers<R> {
      toHaveTimeBetween(expectedMsecs: number): R;
    }
  }
}

expect.extend({
  toHaveTimeBetween([start, end], expectedMsecs: number) {
    const result = getTimeBetween(new Date(start), new Date(end));
    const reverseResult = getTimeBetween(new Date(end), new Date(start));
    const pass = result === expectedMsecs && reverseResult === expectedMsecs;
    return {
      message: () => {
        let msg = '';
        msg += pass ? 'did not expect ' : 'expected ';
        msg += `${expectedMsecs} working milliseconds between "${start}" and "${end}" `;
        msg += `but instead received ${result}`;
        return msg;
      },
      pass
    };
  }
});

const hoursToMs = (n: number) => n * 3600000;
const minsToMs = (n: number) => n * 60000;
const secsToMs = (n: number) => n * 1000;

const getInvalidTimeOptions = (dailyStartOrEnd: string) => [
  { [dailyStartOrEnd]: null },
  { [dailyStartOrEnd]: [] },
  { [dailyStartOrEnd]: [11] },
  { [dailyStartOrEnd]: [11, 0] },
  { [dailyStartOrEnd]: ['11', '00', '00'] },
  { [dailyStartOrEnd]: ['11:00:00'] },
  { [dailyStartOrEnd]: [11, 0, 0, 0] },
  { [dailyStartOrEnd]: [24, 0, 0] },
  { [dailyStartOrEnd]: [0, 60, 0] },
  { [dailyStartOrEnd]: [0, 0, 60] },
  { [dailyStartOrEnd]: [-1, 0, 0] },
  { [dailyStartOrEnd]: [0, -1, 0] },
  { [dailyStartOrEnd]: [0, 0, -1] }
];

[
  ...getInvalidTimeOptions('dailyEnd'),
  ...getInvalidTimeOptions('dailyStart'),
  { excludedDays: null },
  { excludedDays: [null] },
  { excludedDays: ['4'] },
  { excludedDays: [-1] },
  { excludedDays: [7] }
].forEach((opts: any) => {
  it(`should throw when invalid options ${JSON.stringify(
    opts
  )} are provided`, () => {
    expect(() => {
      getTimeBetween(new Date(), new Date(), opts);
    }).toThrow();
  });
});

describe('during week', () => {
  it('should work for single day, overnight', () => {
    const start = 'March 1 2017 12:00:00';
    const end = 'March 2 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(8));
  });

  it('should work regardless of order', () => {
    const start = 'March 1 2017 12:00:00';
    const end = 'March 2 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(8));
  });

  it('should work for single day', () => {
    const start = 'March 2 2017 00:00:00';
    const end = 'March 2 2017 23:59:59';
    expect([start, end]).toHaveTimeBetween(hoursToMs(8));
  });
});

describe('over weekend', () => {
  it('should work for single day', () => {
    const start = 'March 3 2017 12:00:00';
    const end = 'March 6 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(8));
  });

  it('should work for half way into weekend', () => {
    const start = 'March 3 2017 12:00:00';
    const end = 'March 5 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(6));
  });

  it('should work for a week', () => {
    const start = 'March 3 2017 12:00:00';
    const end = 'March 10 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(40));
  });

  it('should work for second weekend', () => {
    const start = 'March 3 2017 12:00:00';
    const end = 'March 11 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(46));
  });

  it('should work for two weekends', () => {
    const start = 'March 3 2017 12:00:00';
    const end = 'March 13 2017 12:00:00';
    expect([start, end]).toHaveTimeBetween(hoursToMs(48));
  });
});

describe('specific bugs', () => {
  it('should work for this odd error case was getting -4', () => {
    const start = '2017-02-18T03:30:00.000Z';
    const end = '2017-02-18T03:30:00.000Z';
    expect([start, end]).toHaveTimeBetween(0);
  });

  it('should work for this odd error case was getting -1', () => {
    const start = '2017-02-13T22:23:19.000Z';
    const end = '2017-02-14T13:51:48.000Z';
    expect([start, end]).toHaveTimeBetween(
      hoursToMs(3) + minsToMs(51) + secsToMs(48)
    );
  });

  it('should work for this odd error case was getting -2', () => {
    const start = '2017-02-14T13:51:48.000Z';
    const end = '2017-02-14T13:51:48.000Z';
    expect([start, end]).toHaveTimeBetween(0);
  });
});


time-between.js
function isSameDate(a, b) {
  return a.getFullYear() === b.getFullYear()
    && a.getMonth() === b.getMonth()
    && a.getDate() === b.getDate()
}

module.exports = function hoursBetween(start, end, opts) {
  if (start.getTime() > end.getTime()) {
    var _end = end
    end = start
    start = _end
  }

  opts || (opts = {})

  var dailyStart = opts.dailyStart || 10
  var dailyEnd = opts.dailyEnd || 18
  var voidDays = opts.voidDays || [0, 6]

  function handleDate(currentDate, start, end) {
    if (voidDays.includes(currentDate.getDay())) {
      return 0
    }

    var isStart = isSameDate(start, currentDate)
    var isEnd = isSameDate(end, currentDate)

    var startHour = isStart ? start.getHours() : 0
    var endHour = isEnd ? end.getHours() : 24

    // console.log({
    //   startHour, dailyStart,
    //   endHour, dailyEnd,
    // })

    startHour = Math.min(Math.max(startHour, dailyStart), dailyEnd)
    endHour = Math.max(Math.min(endHour, dailyEnd), dailyStart)

    // console.log({
    //   start, end,
    //   isStart, isEnd,
    //   startHour, endHour,
    // })

    return endHour - startHour
  }

  var currentDate = new Date(start.getTime())
  currentDate.setHours(0, 0, 0, 0)

  var lastDate = new Date(end.getTime())
  lastDate.setHours(23, 59, 59, 999)
  var endTime = lastDate.getTime()

  var timeElapsed = 0

  while (currentDate.getTime() < endTime) {
    timeElapsed += handleDate(currentDate, start, end)

    currentDate.setDate(currentDate.getDate() + 1)
  }

  return timeElapsed
}


src/time-between.ts
export type Sunday = 0;
export type Monday = 1;
export type Tuesday = 2;
export type Wednesday = 3;
export type Thursday = 4;
export type Friday = 5;
export type Saturday = 6;

export type Day =
  | Sunday
  | Monday
  | Tuesday
  | Wednesday
  | Thursday
  | Friday
  | Saturday;

export type Hour =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23;

interface AllOptions {
  dailyStart: number;
  dailyEnd: number;
  voidDays: number[];
}

const defaults: AllOptions = {
  dailyEnd: 18,
  dailyStart: 10,
  voidDays: [6, 0]
};

export type GetTimeBetween = (
  start: Date,
  end: Date,
  options?: {
    /** Hour of the day that work starts */
    dailyStart?: Hour;
    /** Hour of the day that work ends */
    dailyEnd?: Hour;
    /** Zero-Indexed non-working days */
    voidDays?: Day[];
  }
) => number;

/**
 * Calculate the number of millseconds during working hours between two dates.
 */
export const getTimeBetween: GetTimeBetween = (start, end, options) => {
  const { dailyEnd, dailyStart, voidDays } = {
    ...defaults,
    ...options
  } as AllOptions;
  const [startDate, endDate] =
    start.getTime() > end.getTime() ? [end, start] : [start, end];

  const isSameDate = (a: Date, b: Date) =>
    a.getFullYear() === b.getFullYear() &&
    a.getMonth() === b.getMonth() &&
    a.getDate() === b.getDate();

  const isVoidDay = (date: Date) => voidDays.includes(date.getDay());
  const isOnStartDate = (date: Date) => isSameDate(startDate, date);
  const isOnEndDate = (date: Date) => isSameDate(endDate, date);

  const isBeforeStartOfWorkingHours = (date: Date) =>
    date.getHours() < dailyStart;
  const isAfterEndOfWorkingHours = (date: Date) => date.getHours() > dailyEnd;

  const withTime = (
    date: Date,
    hours: number,
    mins: number,
    secs: number,
    msecs: number
  ) => {
    const nextDate = new Date(date.getTime());
    nextDate.setHours(hours, mins, secs, msecs);
    return nextDate;
  };

  const withTimeAtStartOfDay = (date: Date) => withTime(date, 0, 0, 0, 0);
  const withTimeAtEndOfDay = (date: Date) => withTime(date, 23, 0, 0, 0);
  const withTimeAtStartOfWorkingHours = (date: Date) =>
    withTime(date, dailyStart, 0, 0, 0);
  const withTimeAtEndOfWorkingHours = (date: Date) =>
    withTime(date, dailyEnd, 0, 0, 0);

  const withTimeDuringWorkingHours = (date: Date) => {
    if (isBeforeStartOfWorkingHours(date)) {
      return withTimeAtStartOfWorkingHours(date);
    }
    if (isAfterEndOfWorkingHours(date)) {
      return withTimeAtEndOfWorkingHours(date);
    }
    return date;
  };

  const getStartOfDay = (date: Date) =>
    withTimeDuringWorkingHours(
      isOnStartDate(date) ? startDate : withTimeAtStartOfDay(date)
    );

  const getEndOfDay = (date: Date) =>
    withTimeDuringWorkingHours(
      isOnEndDate(date) ? endDate : withTimeAtEndOfDay(date)
    );

  const nextDay = (date: Date) => {
    const nextDate = new Date(date.getTime());
    nextDate.setDate(date.getDate() + 1);
    return nextDate;
  };

  const startDay = withTimeAtStartOfWorkingHours(startDate);
  const endDay = withTimeAtEndOfWorkingHours(endDate);
  const endTime = endDay.getTime();

  let thisDay = startDay;
  let totalMsecs = 0;

  while (thisDay.getTime() < endTime) {
    if (!isVoidDay(thisDay)) {
      totalMsecs +=
        getEndOfDay(thisDay).getTime() - getStartOfDay(thisDay).getTime();
    }
    thisDay = nextDay(thisDay);
  }

  return totalMsecs;
};

/**
 * Calculate the number of working hours between two dates.
 *
 * Incomplete hours are rounded down – if two dates which are 1hr 45mins apart
 * are given, the result will be 1.
 */
export const getHoursBetween: GetTimeBetween = (start, end, options) => {
  const msecsPerHour = 3600000;
  const msecsBetween = getTimeBetween(start, end, options);
  return Math.floor(msecsBetween / msecsPerHour);
};

export default getHoursBetween;


src/get-time-between.ts
import { Day, Hour, Minute, Second } from './options';

interface Options {
  dailyEnd: [number, number, number];
  dailyStart: [number, number, number];
  excludedDays: number[];
}

export type GetTimeBetween = (
  start: Date,
  end: Date,
  options?: {
    /** Hour of the day that work ends */
    dailyEnd?: [Hour, Minute, Second];
    /** Hour of the day that work starts */
    dailyStart?: [Hour, Minute, Second];
    /** Zero-Indexed non-working days */
    excludedDays?: Day[];
  }
) => number;

/**
 * Calculate the number of millseconds during working hours between two dates.
 */
export const getTimeBetween: GetTimeBetween = (start, end, options) => {
  const isNumber = (n: any) =>
    typeof n === 'number' && !isNaN(n) && isFinite(n);
  const isArrayOfNumbers = (a: any) => Array.isArray(a) && a.every(isNumber);
  const isWithinRange = (floor: number, ceiling: number, n: number) =>
    n >= floor && n <= ceiling;
  const isValidDay = (v: any) => isNumber(v) && isWithinRange(0, 6, v);
  const isArrayOfDays = (a: any) => Array.isArray(a) && a.every(isValidDay);
  const isValidTime = (v: any) =>
    isArrayOfNumbers(v) &&
    v.length === 3 &&
    isWithinRange(0, 23, v[0]) &&
    isWithinRange(0, 59, v[1]) &&
    isWithinRange(0, 59, v[2]);

  const defaults: Options = {
    dailyEnd: [18, 0, 0],
    dailyStart: [10, 0, 0],
    excludedDays: [6, 0]
  };
  const opts = { ...defaults, ...options } as Options;

  if (!isValidTime(opts.dailyStart)) {
    throw new TypeError(
      'dailyStart should be undefined or an Array of 3 Numbers'
    );
  }

  if (!isValidTime(opts.dailyEnd)) {
    throw new TypeError(
      'dailyEnd should be undefined or an Array of 3 Numbers'
    );
  }

  if (!isArrayOfDays(opts.excludedDays)) {
    throw new TypeError(
      'excludedDays should be undefined or an Array of Numbers between 0 (Sunday) and 6 (Saturday)'
    );
  }

  const { dailyEnd, dailyStart, excludedDays } = opts;
  const [startDate, endDate] =
    start.getTime() > end.getTime() ? [end, start] : [start, end];

  const isSameDate = (a: Date, b: Date) =>
    a.getFullYear() === b.getFullYear() &&
    a.getMonth() === b.getMonth() &&
    a.getDate() === b.getDate();

  const isVoidDay = (date: Date) => excludedDays.includes(date.getDay());
  const isOnStartDate = (date: Date) => isSameDate(startDate, date);
  const isOnEndDate = (date: Date) => isSameDate(endDate, date);

  const isBeforeStartOfWorkingHours = (date: Date) =>
    date.getHours() < dailyStart[0];
  const isAfterEndOfWorkingHours = (date: Date) =>
    date.getHours() > dailyEnd[0];

  const withTime = (
    date: Date,
    hours: number,
    mins: number,
    secs: number,
    msecs: number
  ) => {
    const nextDate = new Date(date.getTime());
    nextDate.setHours(hours, mins, secs, msecs);
    return nextDate;
  };

  const withTimeAtStartOfDay = (date: Date) => withTime(date, 0, 0, 0, 0);
  const withTimeAtEndOfDay = (date: Date) => withTime(date, 23, 0, 0, 0);
  const withTimeAtStartOfWorkingHours = (date: Date) =>
    withTime(date, dailyStart[0], dailyStart[1], dailyStart[2], 0);
  const withTimeAtEndOfWorkingHours = (date: Date) =>
    withTime(date, dailyEnd[0], dailyEnd[1], dailyEnd[2], 0);

  const withTimeDuringWorkingHours = (date: Date) => {
    if (isBeforeStartOfWorkingHours(date)) {
      return withTimeAtStartOfWorkingHours(date);
    }
    if (isAfterEndOfWorkingHours(date)) {
      return withTimeAtEndOfWorkingHours(date);
    }
    return date;
  };

  const getStartOfDay = (date: Date) =>
    withTimeDuringWorkingHours(
      isOnStartDate(date) ? startDate : withTimeAtStartOfDay(date)
    );

  const getEndOfDay = (date: Date) =>
    withTimeDuringWorkingHours(
      isOnEndDate(date) ? endDate : withTimeAtEndOfDay(date)
    );

  const nextDay = (date: Date) => {
    const nextDate = new Date(date.getTime());
    nextDate.setDate(date.getDate() + 1);
    return nextDate;
  };

  const startDay = withTimeAtStartOfWorkingHours(startDate);
  const endDay = withTimeAtEndOfWorkingHours(endDate);
  const endTime = endDay.getTime();

  let thisDay = startDay;
  let totalMsecs = 0;

  while (thisDay.getTime() < endTime) {
    if (!isVoidDay(thisDay)) {
      totalMsecs +=
        getEndOfDay(thisDay).getTime() - getStartOfDay(thisDay).getTime();
    }
    thisDay = nextDay(thisDay);
  }

  return totalMsecs;
};


src/options.ts
export type Sunday = 0;
export type Monday = 1;
export type Tuesday = 2;
export type Wednesday = 3;
export type Thursday = 4;
export type Friday = 5;
export type Saturday = 6;

export type Day =
  | Sunday
  | Monday
  | Tuesday
  | Wednesday
  | Thursday
  | Friday
  | Saturday;

export type Hour =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23;

export type Minute =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23
  | 24
  | 25
  | 26
  | 27
  | 28
  | 29
  | 30
  | 31
  | 32
  | 33
  | 34
  | 35
  | 36
  | 37
  | 38
  | 39
  | 40
  | 40
  | 41
  | 42
  | 43
  | 44
  | 45
  | 46
  | 47
  | 48
  | 49
  | 50
  | 50
  | 51
  | 52
  | 53
  | 54
  | 55
  | 56
  | 57
  | 58
  | 59;

export type Second = Minute;
